{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gencode.vM1.pc_transcripts.fa.gz\n",
      "55442\n",
      "mmu_mm9_gencode_vM1.transcripts.fa.gz\n",
      "26589\n",
      "=== Stats for mouse (Gencode M1)\n",
      "Counter({'match': 26579, 'no-key': 10})\n",
      "gencode.v19.pc_transcripts.fa.gz\n",
      "95308\n",
      "hsa_hg19_gencode_v19.transcripts.fa.gz\n"
     ]
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import os as os\n",
    "import gzip as gz\n",
    "import collections as col\n",
    "\n",
    "# What is this?\n",
    "# Just make sure that the filter_genemodel.py script produces the same output\n",
    "# as can be readily downloaded from GENCODE for human and mouse\n",
    "# (modulo minor discrepancies depending on definition of protein coding gene)\n",
    "\n",
    "genc_dir = '/TL/deep/fhgfs/projects/pebert/thesis/refdata/transcriptome/rawdata'\n",
    "my_dir = '/TL/deep/fhgfs/projects/pebert/thesis/refdata/transcriptome/fasta'\n",
    "\n",
    "def read_transcripts(fpath):\n",
    "    collect = dict()\n",
    "    tname = ''\n",
    "    tseq = ''\n",
    "    print(os.path.basename(fpath))\n",
    "    with gz.open(fpath, 'rt') as infile:\n",
    "        for line in infile:\n",
    "            if not line.strip():\n",
    "                continue\n",
    "            if line.startswith('>'):\n",
    "                if tname:\n",
    "                    collect[tname.strip()] = tseq.upper()\n",
    "                tseq = ''\n",
    "                new_name = line.strip('>').split('|')[0].split('.')[0]\n",
    "                tname = new_name\n",
    "            else:\n",
    "                tseq += line.strip()\n",
    "    print(len(collect))\n",
    "    return collect\n",
    "\n",
    "genc_mmu = read_transcripts(os.path.join(genc_dir, 'gencode.vM1.pc_transcripts.fa.gz'))\n",
    "\n",
    "my_mmu = read_transcripts(os.path.join(my_dir, 'mmu_mm9_gencode_vM1.transcripts.fa.gz'))\n",
    "\n",
    "stats = col.Counter()\n",
    "for k, seq in my_mmu.items():\n",
    "    try:\n",
    "        genc_seq = genc_mmu[k]\n",
    "        if genc_seq == seq:\n",
    "            stats['match'] += 1\n",
    "        else:\n",
    "            stats['no-match'] += 1\n",
    "    except KeyError:\n",
    "        stats['no-key'] += 1\n",
    "print('=== Stats for mouse (Gencode M1)')\n",
    "print(stats)\n",
    "        \n",
    "genc_hsa = read_transcripts(os.path.join(genc_dir, 'gencode.v19.pc_transcripts.fa.gz'))\n",
    "\n",
    "my_hsa = read_transcripts(os.path.join(my_dir, 'hsa_hg19_gencode_v19.transcripts.fa.gz'))\n",
    "\n",
    "stats = col.Counter()\n",
    "for k, seq in my_hsa.items():\n",
    "    try:\n",
    "        genc_seq = genc_hsa[k]\n",
    "        if genc_seq == seq:\n",
    "            stats['match'] += 1\n",
    "        else:\n",
    "            stats['no-match'] += 1\n",
    "    except KeyError:\n",
    "        stats['no-key'] += 1\n",
    "print('=== Stats for human (Gencode 19)')\n",
    "print(stats)\n",
    "\n",
    "\n",
    "\n",
    "    \n"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
