{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing  hg19_from_mm9\n",
      "Writing  hg19_ncd4_top-marked-genes_promoter.bed\n",
      "Writing  hg19_ncd4_top-marked-genes_body.bed\n",
      "Writing  hg19_esc_top-marked-genes_promoter.bed\n",
      "Writing  hg19_esc_top-marked-genes_body.bed\n",
      "Writing  hg19_liver_top-marked-genes_promoter.bed\n",
      "Writing  hg19_liver_top-marked-genes_body.bed\n",
      "Writing  hg19_heart_top-marked-genes_promoter.bed\n",
      "Writing  hg19_heart_top-marked-genes_body.bed\n",
      "Writing  hg19_kidney_top-marked-genes_promoter.bed\n",
      "Writing  hg19_kidney_top-marked-genes_body.bed\n",
      "Processing  mm9_from_hg19\n",
      "Writing  mm9_esc_top-marked-genes_promoter.bed\n",
      "Writing  mm9_esc_top-marked-genes_body.bed\n",
      "Writing  mm9_hepa_top-marked-genes_promoter.bed\n",
      "Writing  mm9_hepa_top-marked-genes_body.bed\n",
      "Writing  mm9_ncd4_top-marked-genes_promoter.bed\n",
      "Writing  mm9_ncd4_top-marked-genes_body.bed\n",
      "Done\n"
     ]
    }
   ],
   "source": [
    "\n",
    "import os as os\n",
    "import collections as col\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# What is this?\n",
    "# Extract sets of gene promoters that show a strong\n",
    "# histone signal (weighted by alignment rate) after\n",
    "# cross-species transfer of the epigenetic signal\n",
    "#\n",
    "# Dump that list as BED file for a LOLA enrichment analysis\n",
    "\n",
    "project_folder = '/TL/deep/fhgfs/projects/pebert/thesis/projects/cross_species/processing/norm'\n",
    "\n",
    "testdata_root = os.path.join(project_folder, 'task_testdata_exp/test_datasets')\n",
    "\n",
    "output_folder = os.path.join(project_folder, 'task_lola', 'top_marked_genes')\n",
    "os.makedirs(output_folder, exist_ok=True)\n",
    "\n",
    "done_epigenomes = set()\n",
    "\n",
    "select_cols =  ['chrom', 'start_body', 'end_body', 'strand_body',\n",
    "                'start_reg5p', 'end_reg5p', 'strand_reg5p',\n",
    "                'name', 'symbol',\n",
    "                'ftmsig_H3K36me3_pct_cons_body', 'ftmsig_H3K36me3_abs_mean_body',\n",
    "                'ftmsig_H3K4me3_pct_cons_reg5p', 'ftmsig_H3K4me3_abs_mean_reg5p']\n",
    "prom_cols = ['chrom', 'start_reg5p', 'end_reg5p', 'strand_reg5p', 'name', 'symbol', 'score_promoter']\n",
    "body_cols = ['chrom', 'start_body', 'end_body', 'strand_body', 'name', 'symbol', 'score_gene']\n",
    "\n",
    "for root, dirs, featfiles in os.walk(testdata_root):\n",
    "    if featfiles:\n",
    "        gene_collector = None\n",
    "        qry, _ ,trg = os.path.split(root)[-1].split('_')\n",
    "        if (trg, qry) not in [('hg19', 'mm9'), ('mm9', 'hg19')]:\n",
    "            continue\n",
    "        print('Processing ', os.path.split(root)[-1])\n",
    "        for ff in featfiles:\n",
    "            if ff.startswith('G9930'):\n",
    "                continue\n",
    "            _, epigenome, _ = ff.split('_', 2)\n",
    "            if epigenome in done_epigenomes:\n",
    "                continue\n",
    "            tissue = ff.split('.')[0].split('_')[-1]\n",
    "            if tissue in ['H1hESC', 'ESE14']:\n",
    "                tissue = 'esc'\n",
    "            fpath = os.path.join(root, ff)\n",
    "            dataset = []\n",
    "            with pd.HDFStore(fpath, 'r') as hdf:\n",
    "                for k in hdf.keys():\n",
    "                    if k.startswith('/metadata'):\n",
    "                        continue\n",
    "                    chrom = os.path.split(k)[-1]\n",
    "                    data = hdf[k]\n",
    "                    data['chrom'] = chrom\n",
    "                    dataset.append(data)\n",
    "            dataset = pd.concat(dataset, axis=0, ignore_index=False)\n",
    "            dataset = dataset.loc[:, select_cols].copy()\n",
    "            dataset['tissue'] = tissue\n",
    "            #dataset['score_promoter'] = dataset['ftmsig_H3K4me3_abs_mean_reg5p'] * (dataset['ftmsig_H3K4me3_pct_cons_reg5p'] / 100)\n",
    "            dataset['score_promoter'] = dataset['ftmsig_H3K4me3_abs_mean_reg5p']\n",
    "\n",
    "            dataset['rank_promoter'] = dataset['score_promoter'].rank(pct=True)\n",
    "            \n",
    "            #dataset['score_gene'] = dataset['ftmsig_H3K36me3_abs_mean_body'] * (dataset['ftmsig_H3K36me3_pct_cons_body'] / 100)\n",
    "            dataset['score_gene'] = dataset['ftmsig_H3K36me3_abs_mean_body']\n",
    "\n",
    "            dataset['rank_gene'] = dataset['score_gene'].rank(pct=True)\n",
    "            \n",
    "            # ranking: select >= 0.95 for output to select top 5%\n",
    "            t = 0.95\n",
    "            \n",
    "            select_genes = np.logical_or(dataset['rank_promoter'] >= t, dataset['rank_gene'] >= t)\n",
    "            dataset = dataset.loc[select_genes, :].copy()\n",
    "            if gene_collector is None:\n",
    "                gene_collector = dataset.copy()\n",
    "            else:\n",
    "                gene_collector = pd.concat([gene_collector, dataset], axis=0, ignore_index=False)\n",
    "            done_epigenomes.add(epigenome)\n",
    "            \n",
    "        # dump\n",
    "        for tissue in gene_collector['tissue'].unique():\n",
    "            sub_select = np.logical_and(gene_collector['tissue'] == tissue,\n",
    "                                        gene_collector['rank_promoter'] >= t)\n",
    "            prom_subset = gene_collector.loc[sub_select, prom_cols]\n",
    "            mean_score = prom_subset.loc[:, ['name', 'score_promoter']].groupby('name').mean()\n",
    "            prom_subset.drop_duplicates(['name', 'symbol'], inplace=True)\n",
    "            prom_subset.index = prom_subset['name']\n",
    "            prom_subset['mean_score'] = mean_score\n",
    "            prom_subset.sort_values(['chrom', 'start_reg5p', 'end_reg5p'], ascending=True, inplace=True)\n",
    "            \n",
    "            outfile = '{}_{}_top-marked-genes_promoter.bed'.format(qry, tissue)\n",
    "            print('Writing ', outfile)\n",
    "            outpath = os.path.join(output_folder, outfile)\n",
    "            prom_subset.to_csv(outpath, sep='\\t', index=False, index_label=False, mode='w', header=False,\n",
    "                              columns=['chrom', 'start_reg5p', 'end_reg5p', 'name', 'mean_score',\n",
    "                                       'strand_reg5p', 'symbol'])\n",
    "            \n",
    "            # same for gene bodies\n",
    "            sub_select = np.logical_and(gene_collector['tissue'] == tissue,\n",
    "                                        gene_collector['rank_gene'] >= t)\n",
    "            body_subset = gene_collector.loc[sub_select, body_cols]\n",
    "            mean_score = body_subset.loc[:, ['name', 'score_gene']].groupby('name').mean()\n",
    "            body_subset.drop_duplicates(['name', 'symbol'], inplace=True)\n",
    "            body_subset.index = body_subset['name']\n",
    "            body_subset['mean_score'] = mean_score\n",
    "            body_subset.sort_values(['chrom', 'start_body', 'end_body'], ascending=True, inplace=True)\n",
    "            \n",
    "            outfile = '{}_{}_top-marked-genes_body.bed'.format(qry, tissue)\n",
    "            print('Writing ', outfile)\n",
    "            outpath = os.path.join(output_folder, outfile)\n",
    "            body_subset.to_csv(outpath, sep='\\t', index=False, index_label=False, mode='w', header=False,\n",
    "                              columns=['chrom', 'start_body', 'end_body', 'name', 'mean_score',\n",
    "                                       'strand_body', 'symbol'])\n",
    "print('Done')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
